pkg/catalogv2/git/download.go

```diff
@@ -1,124 +1,125 @@
 package git
 
 import (
-	"crypto/sha256"
-	"encoding/hex"
-	"encoding/pem"
 	"fmt"
-	"net/url"
-	"os"
-	"path/filepath"
-	"regexp"
-	"strings"
 
 	"github.com/rancher/rancher/pkg/settings"
 	corev1 "k8s.io/api/core/v1"
 )
 
-const (
-	stateDir  = "management-state/git-repo"
-	staticDir = "/var/lib/rancher-data/local-catalogs/v2"
-	localDir  = "../rancher-data/local-catalogs/v2" // identical to helm.InternalCatalog
-)
+// Ensure builds the configuration for a should-existing repo and makes sure it is cloned or reseted to the latest commit of given branch
+func Ensure(secret *corev1.Secret, namespace, name, gitURL, branch string, insecureSkipTLS bool, caBundle []byte) error {
+	git, err := gitForRepo(secret, namespace, name, gitURL, insecureSkipTLS, caBundle)
+	if err != nil {
+		return err
+	}
 
-func gitDir(namespace, name, gitURL string) string {
-	staticDir := filepath.Join(staticDir, namespace, name, hash(gitURL))
-	if s, err := os.Stat(staticDir); err == nil && s.IsDir() {
-		return staticDir
+	return git.EnsureClonedRepo(branch)
+}
+
+// EnsureClonedRepo will check if repo is cloned, if not the method will clone and reset to the latest commit.
+// If reseting to the latest commit is not possible it will fetch and try to reset again
+func (er *extendedRepo) EnsureClonedRepo(branch string) error {
```

Comment:
Where is the extendedRepo struct in here ? I think the struct should be in this file since the functions of the struct are in this file

Reply:
Okay, why do we need an interface repository and two structs repo operations and extendedRepo structs here which implements repository interface? 

We could have just done something like 

```
type git struct {

}

func newGit() {
  return Git{}
}

func(g *git) CloneOrOpen() {}

func(g *git) HardReset() {}

func(g *git) FetchAndReset() {}

....
```

The Ensure, Head, and Update functions can create a new Git client using the gitForRepo and then use that git to call these functions.

Let me know if I am missing something here. One Struct ie. Git seems sufficient here instead of an interface and two structs. We can write unit tests for the Git struct as well?

Reply:
I have implemented a fix, please resolve the conversation